### PR TITLE
Add CMake build system

### PR DESCRIPTION
**Pull request by ProExpertProg**
Added a CMake-based build system to enable building this package from source in the vLLM build. That will allow using PyTorch nightly, which is needed by the `torch.compile` and another project in vLLM.

Heavily inspired by the [new build system in vLLM](https://github.com/vllm-project/vllm/pull/2830).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,239 +1,233 @@
...
+        # If we've already configured using the CMakeLists.txt for
+        # this extension, exit early.
+        if ext.cmake_lists_dir in cmake_build_ext.did_config:
+            return
+
+        cmake_build_ext.did_config[ext.cmake_lists_dir] = True
+
+        # Select the build type.
+        # Note: optimization level + debug info are set by the build type
+        default_cfg = "Debug" if self.debug else "RelWithDebInfo"
```
Comment (bnellnm) [inline setup.py:144]: Can you double check that CMake uses -O3 for the optimization level with the config type is Release or RelWithDebInfo?  Just so this matches the existing behavior.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+find_library(TORCH_PYTHON_LIBRARY torch_python PATHS "${TORCH_INSTALL_PREFIX}/lib" REQUIRED)
+target_link_libraries(_C PRIVATE ${TORCH_PYTHON_LIBRARY})
+
+target_include_directories(_C PRIVATE
+        csrc/flash_attn
+        csrc/flash_attn/src
+        csrc/cutlass/include)
+
+# remove macros
+target_compile_options(_C PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__>)
```
Comment (bnellnm) [inline CMakeLists.txt:151]: `get_torch_gpu_compiler_flags()` should remove all these already.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+# other options
+target_compile_options(_C PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr --expt-extended-lambda --use_fast_math>)
+# custom definitions (TODO: both langs?)
+target_compile_definitions(_C PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:FLASHATTENTION_DISABLE_DROPOUT FLASHATTENTION_DISABLE_UNEVEN_K>)
```
Comment (bnellnm) [inline CMakeLists.txt:155]: You should be able to set these with `COMPILE_FLAGS` on `define_gpu_extension_target`

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+# TODO(luka): remove this once PyBind is removed
+find_library(TORCH_PYTHON_LIBRARY torch_python PATHS "${TORCH_INSTALL_PREFIX}/lib" REQUIRED)
+target_link_libraries(_C PRIVATE ${TORCH_PYTHON_LIBRARY})
+
+target_include_directories(_C PRIVATE
+        csrc/flash_attn
+        csrc/flash_attn/src
+        csrc/cutlass/include)
```
Comment (bnellnm) [inline CMakeLists.txt:148]: You could use `INCLUDE_DIRECTORIES` on `define_gpu_extension_target` to set these.

File: setup.py [hunk 5]
```diff
@@ -1,239 +1,233 @@
...
+ext_modules = []
 
-def get_cuda_bare_metal_version(cuda_dir):
-    raw_output = subprocess.check_output([cuda_dir + "/bin/nvcc", "-V"], universal_newlines=True)
-    output = raw_output.split()
-    release_idx = output.index("release") + 1
-    bare_metal_version = parse(output[release_idx].split(",")[0])
+# We want this even if SKIP_CUDA_BUILD because when we run python setup.py sdist we want the .hpp
+# files included in the source distribution, in case the user compiles from source.
+subprocess.run(["git", "submodule", "update", "--init", "csrc/cutlass"])
```
Comment (bnellnm) [inline setup.py:49]: Maybe a TODO for later but we could replace this with the same FetchContent method as the vllm CMakeLists?

File: setup.py (hunk 1 above)
Comment (ProExpertProg) [reply setup.py:144]: `RelWithDebInfo` is `-O2` by default and `Release` is `-O3`. Should we just always append `-O3`?

Comment (bnellnm) [reply setup.py:144]: Yeah, or change the default build type to `Release`?

File: CMakeLists.txt (hunk 4 above)
Comment (ProExpertProg) [reply CMakeLists.txt:148]: To me this is clearer, and it doesn't require redundant proxy variables, and doesn't duplicate functionality (and is `INCLUDE_DIRECTORIES` on the GPU target `PRIVATE` or `PUBLIC` by default)? I'm happy to change it though if you have a strong preference. I switched the other flags over to avoid repeating the generator expression

Comment (bnellnm) [reply CMakeLists.txt]: It defaults to `PRIVATE`.  I don't have super strong feelings about either style.

File: csrc/flash_attn/flash_api.cpp [hunk 6]
```diff
@@ -17,9 +18,13 @@
 #define CHECK_SHAPE(x, ...) TORCH_CHECK(x.sizes() == torch::IntArrayRef({__VA_ARGS__}), #x " must have shape (" #__VA_ARGS__ ")")
 #define CHECK_CONTIGUOUS(x) TORCH_CHECK(x.is_contiguous(), #x " must be contiguous")
 
+// Tuple with N elements of the same type
+template <typename T, size_t N>
+using Tuple = void;
+
```
Comment (bnellnm) [inline csrc/flash_attn/flash_api.cpp:24]: What is this for?

Comment (ProExpertProg) [reply csrc/flash_attn/flash_api.cpp:24]: Oops I meant to resolve this

File: csrc/flash_attn/flash_api.cpp [hunk 7]
```diff
@@ -972,11 +1002,26 @@ mha_fwd_kvcache(at::Tensor &q,                 // batch_size x seqlen_q x num_he
...
+    ops.def("varlen_fwd(Tensor! q, Tensor k, Tensor v, Tensor!? out, Tensor cu_seqlens_q, "
+            "Tensor cu_seqlens_k, Tensor? seqused_k, Tensor? block_table, Tensor? alibi_slopes, "
+            "int max_seqlen_q, int max_seqlen_k, float p_dropout, float softmax_scale, bool zero_tensors, "
+            "bool is_causal, int window_size_left, int window_size_right, float softcap, bool return_softmax, "
+            "Generator? gen) -> Tensor[]");
+    ops.impl("varlen_fwd", torch::kCUDA, &mha_varlen_fwd);
+
+    ops.def("fwd_kvcache(Tensor! q, Tensor kcache, Tensor vcache, Tensor? k, Tensor? v, Tensor? seqlens_k, "
+            "Tensor? rotary_cos, Tensor? rotary_sin, Tensor? cache_batch_idx, Tensor? block_table, Tensor? alibi_slopes, "
+            "Tensor? out, float softmax_scale, bool is_causal, int window_size_left, int window_size_right, "
```
Comment (bnellnm) [inline csrc/flash_attn/flash_api.cpp:1022]: is this `out` a mutable output argument?

File: tests/flash_attn_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+from typing import Optional, List, Tuple
```
Comment (bnellnm) [inline tests/flash_attn_wrapper.py:1]: Is this file necessary now that you've added torch bindings?

Comment (ProExpertProg) [reply tests/flash_attn_wrapper.py:1]: I added it more because I wanted to keep the test intact so it can be kept in sync with the vllm file. But yeah the wrapper is not necessary

File: CMakeLists.txt (hunk 4 above)
Comment (ProExpertProg) [reply CMakeLists.txt]: Yeah it was more of a rhetorical question 😅 . I'll keep this one

File: csrc/flash_attn/flash_api.cpp [hunk 9]
```diff
@@ -255,18 +262,18 @@ void set_params_alibi(Flash_fwd_params &params, c10::optional<at::Tensor> &alibi
 #endif
 }
 
-std::vector<at::Tensor>
+std::array<at::Tensor, 8>
```
Comment (bnellnm) [inline csrc/flash_attn/flash_api.cpp:265]: Why change this to an array?

File: csrc/flash_attn/flash_api.cpp [hunk 10]
```diff
@@ -442,25 +453,25 @@ mha_fwd(at::Tensor &q,         // batch_size x seqlen_q x num_heads x head_size
     return {out, q_padded, k_padded, v_padded, out_padded, softmax_lse, p, rng_state};
 }
 
-std::vector<at::Tensor>
+std::array<at::Tensor, 8>
```
Comment (bnellnm) [inline csrc/flash_attn/flash_api.cpp:456]: Ditto.

File: csrc/flash_attn/flash_api.cpp [hunk 11]
```diff
@@ -695,27 +713,27 @@ mha_varlen_fwd(at::Tensor &q,  // total_q x num_heads x head_size, total_q := \s
     return {out, q_padded, k_padded, v_padded, out_padded, softmax_lse, p, rng_state};
 }
 
-std::vector<at::Tensor>
+std::array<at::Tensor, 2>
```
Comment (bnellnm) [inline csrc/flash_attn/flash_api.cpp:716]: Ditto

File: csrc/flash_attn/flash_api.cpp
Comment (bnellnm) [inline csrc/flash_attn/flash_api.cpp:1]: I know the original formatting was not great but it would be a bit easier to review if the only changes were related to torch registration.

File: csrc/flash_attn/flash_api.cpp (hunk 9 above)
Comment (ProExpertProg) [reply csrc/flash_attn/flash_api.cpp:265]: If I remember correctly, it didn't properly work with torch extension but I can try again

Comment (bnellnm) [reply csrc/flash_attn/flash_api.cpp:265]: `std::vector<at::Tensor>` should work.  The schema type would need to be `Tensor[]`.

File: tests/test_vllm_flash_attn.py
Comment (bnellnm) [inline tests/test_vllm_flash_attn.py:1]: Can you add some [torch.library.opcheck](https://pytorch.org/docs/stable/library.html#torch.library.opcheck) tests to this file?

It should be pretty simple to do.  opcheck makes copies of all the arguments so you should not need to worry about it trashing any in/out args.

Comment (ProExpertProg) [reply tests/test_vllm_flash_attn.py:1]: It's funny because I actually removed these manually when copying over the test. But I can just add them back in

File: tests/test_vllm_flash_attn.py [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+        causal=True,
+        block_table=block_tables,
+        cache_seqlens=kv_lens_tensor,
+        softcap=soft_cap if soft_cap is not None else 0,
+    ).squeeze(1)
+
+    if num_blocks <= 2048:
+        test_utils = ["test_faketensor", "test_schema"]
+    else:
+        test_utils = ["test_faketensor"]
```
Comment (bnellnm) [inline tests/test_vllm_flash_attn.py:127]: Can you add a comment explaining why `test_schema` isn't used when `num_blocks` > 2048?

File: tests/test_vllm_flash_attn.py [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+        causal=True,
+        window_size=window_size,
+        block_table=block_tables,
+        softcap=soft_cap if soft_cap is not None else 0,
+    )
+
+    if num_blocks <= 2048:
+        test_utils = ["test_faketensor", "test_schema"]
+    else:
+        test_utils = ["test_faketensor"]
```
Comment (bnellnm) [inline tests/test_vllm_flash_attn.py:232]: ditto

**Review by bnellnm — APPROVED** [review]
LGTM.  Just one minor comment in the test file.

**Review by bnellnm — CHANGES_REQUESTED** [review]
Sorry, I missed the python finding code first time.

File: cmake/utils.cmake [hunk 14]
```diff
@@ -0,0 +1,369 @@
+# This file is taken from github.com/vllm-project/vllm/cmake/utils.cmake
+# It contains utility functions for building PyTorch extensions with GPU support.
+# The only modification is finding the `Python` package with a specific version.
+#
+# THIS MACRO WAS MODIFIED
+# Attempt to find the python package and verifies that it is one of the `SUPPORTED_VERSIONS`.
+# To customize which python gets found, set the `Python_EXECUTABLE` variable.
+macro (find_python_constrained_versions SUPPORTED_VERSIONS)
```
Comment (bnellnm) [inline cmake/utils.cmake:8]: The method of finding the python version in vLLM by using the executable was intentional to avoid ambiguities on systems with multiple python versions.  I think we should use the same logic here to prevent any surprises.

Comment (ProExpertProg) [reply cmake/utils.cmake:8]: The logic has not changed. This just slightly simplifies the logic:
- before, if you had one clear python version or not, you had to manually specify it.
- now, you either have a default that _just works_, or you manually specify if you have to. And `setup.py` always specifies manually

So the only change is for people directly invoking CMake from the terminal that don't have ambiguous Python versions - something I do.

Comment (bnellnm) [reply cmake/utils.cmake:8]: I don't see how the logic is the same here.  The old logic intentionally used the exact version derived from `VLLM_PYTHON_EXECUTABLE`.  Afaik `find_package` will just pick the first version of python it can find.

This may not matter much for local development but it matters for CI and packaging,

Comment (ProExpertProg) [reply cmake/utils.cmake:8]: The way I understand this is that `find_package` will find the first Python it can find (which is good in the CMake-from-terminal case) if and only if `Python_EXECUTABLE` is not specified. If `Python_EXECUTABLE` is specified (in the same way `VLLM_PYTHON_EXECUTABLE` was, using `-D`), that version is found. For what it's worth, the old logic manually set `Python_EXECUTABLE` from `VLLM_PYTHON_EXECUTABLE` so this new logic just skips that assignment. And the CI and packaging use `pip install ...` which uses `setup.py` which will always set `Python_EXECUTABLE` anyway. So I believe those cases shouldn't be affected? Am I missing something?

Comment (bnellnm) [reply cmake/utils.cmake:8]: I didn't see where `Python_EXECUTABLE` was being set?  setup.py still sets `VLLM_PYTHON_EXECUTABLE`.  Am I missing something?

File: tests/test_vllm_flash_attn.py (hunk 13 above)
Comment (ProExpertProg) [reply tests/test_vllm_flash_attn.py:237]: Copied verbatim, added comment

File: tests/test_vllm_flash_attn.py (hunk 12 above)
Comment (ProExpertProg) [reply tests/test_vllm_flash_attn.py:132]: Copied verbatim, added comment

File: cmake/utils.cmake (hunk 14 above)
Comment (ProExpertProg) [reply cmake/utils.cmake:8]: Ohh I see good catch - fixed!

**Review by bnellnm — APPROVED** [review]

**Comment by ProExpertProg** [comment]
(marking as draft until I resolve the binary size/arch explosion)

**Comment by ProExpertProg** [comment]
Closed in favor of #19

**Comment by ProExpertProg** [comment]
Closed in favor of #19
